WcaOnRails/app/webpacker/components/CompetitionOverview/CompetitionFilters.js

```diff
@@ -0,0 +1,662 @@
+      if (lastPage.data.length < COMPETITIONS_API_PAGINATION) {
+        return undefined;
+      }
+
+      return allPages.length + 1;
+    },
+  });
+
+  useEffect(() => {
+    const flatData = competitionsData?.pages
+      .map((page) => page.data)
+      .flatMap((comp) => comp);
+
+    if (!flatData) return;
+
+    if (timeOrder === 'present') {
+      const inProgressData = flatData.filter((comp) => comp.inProgress);
+      const notInProgressData = flatData.filter((comp) => !comp.inProgress);
+      setInProgressComps(inProgressData);
+      setNotInProgressFutureComps(notInProgressData);
+    } else if (timeOrder === 'recent') {
+      setRecentComps(flatData);
+    } else if (timeOrder === 'past') {
+      editPastComps(flatData, pastSelectedYear);
+    } else if (timeOrder === 'by_announcement') {
+      setSortByAnnouncementComps(flatData);
+    } else if (timeOrder === 'custom') {
+      setCustomDatesComps(flatData);
+    }
+
+    if (displayMode === 'map') {
+      if (timeOrder === 'present') {
+        setMapDisplayComps(flatData.filter((comp) => !comp.inProgress));
+      } else {
+        setMapDisplayComps(flatData);
+      }
+    }
+  }, [competitionsData, timeOrder, pastSelectedYear, displayMode]);
+
+  const { ref, inView: bottomInView } = useInView();
+  useEffect(() => {
+    if (bottomInView) {
+      competitionsFetchNextPage();
+    }
+  }, [bottomInView, competitionsFetchNextPage]);
+  useEffect(() => {
+    if (hasMoreCompsToLoad && displayMode === 'map' && mapDisplayComps?.length < MAP_DISPLAY_LIMIT) {
+      competitionsFetchNextPage();
+    }
+  }, [competitionsData, displayMode, hasMoreCompsToLoad, mapDisplayComps,
+    competitionsFetchNextPage]);
+
+  const [delegatesInfo, setDelegatesInfo] = useState([]);
+  const {
+    data: delegatesData,
+    fetchNextPage: delegateFetchNextPage,
+    hasNextPage: delegateHasNextPage,
+  } = useInfiniteQuery({
+    queryKey: ['delegates'],
+    queryFn: ({ pageParam = 1 }) => fetchJsonOrError(`${delegatesApiUrl}?page=${pageParam}`),
+    getNextPageParam: (lastPage, allPages) => {
+      // Assuming the last page has less than the max number of competitions fetched per query
+      if (lastPage.data.length < COMPETITIONS_API_PAGINATION) {
+        return undefined;
+      }
+
+      return allPages.length + 1;
+    },
+  });
+  useEffect(() => {
+    const flatData = delegatesData?.pages
+      .map((page) => page.data)
+      .flatMap((delegate) => delegate);
+    setDelegatesInfo(flatData);
+
+    if (delegateHasNextPage) {
+      delegateFetchNextPage();
+    }
+  }, [delegatesData, delegateHasNextPage, delegateFetchNextPage]);
+
+  const customTimeSelectionButton = (
+    <Button
+      primary
+      name="state"
+      id="custom"
+      value="custom"
+      onClick={() => editTimeOrder('custom')}
+      active={timeOrder === 'custom'}
+    >
+      <span className="caption">{I18n.t('competitions.index.custom')}</span>
+    </Button>
+  );
+
+  return (
```

Comment:
So this component is called `CompetitionFilters`, but it renders both the filters and the resulting data, which is confusing.

You can refactor out the filters and the actual data into separate components. And then the filters themselves can also be refactored into separate components per filter. All that will drastically simplify the return body here. It might seem like a lot of prop-drilling will happen (deep chains of components that are just passing props along without using them) but I don't think that will be the case because of the `filterState` object I suggested (switching to a reducer) and the single `reduceFilter` function which replaces all the `edit...` functions.

Reply:
Nice, looking a lot better now. The return statement is still quite long (over 200 lines) so breaking out a separate component for each filter UI element might be worthwhile (up to you).

Reply:
Yep, was going to, cleaned up now!